File: backend/app/schemas/document.py
from datetime import datetime
from pydantic import BaseModel, Field

from app.models.document import DocumentStatus, DocumentType


class DocumentCreate(BaseModel):
    title: str
    document_type: DocumentType = DocumentType.other
    description: str | None = None
    status: DocumentStatus = DocumentStatus.draft
    course_id: int | None = None
    class_offering_id: int | None = None
    organization_id: int | None = None
    student_id: int | None = None
    external_reference: str | None = None


class DocumentUpdate(BaseModel):
    title: str | None = None
    document_type: DocumentType | None = None
    description: str | None = None
    status: DocumentStatus | None = None
    course_id: int | None = None
    class_offering_id: int | None = None
    organization_id: int | None = None
    student_id: int | None = None
    external_reference: str | None = None
    is_signed: bool | None = None
    signed_at: datetime | None = None
    signed_by: str | None = None


class DocumentVersionCreate(BaseModel):
    notes: str | None = None
    external_url: str | None = None


class DocumentVersionOut(BaseModel):
    id: int
    document_id: int
    version_number: int
    file_name: str | None
    mime_type: str | None
    file_size: int | None
    external_url: str | None
    notes: str | None
    created_by_id: int | None
    created_at: datetime

    model_config = {"from_attributes": True}


class DocumentOut(BaseModel):
    id: int
    title: str
    document_type: DocumentType
    description: str | None
    status: DocumentStatus
    course_id: int | None
    class_offering_id: int | None
    organization_id: int | None
    student_id: int | None
    uploaded_by_id: int | None
    latest_version_number: int
    is_signed: bool
    signed_at: datetime | None
    signed_by: str | None
    external_reference: str | None
    created_at: datetime
    updated_at: datetime
    versions: list[DocumentVersionOut] = Field(default_factory=list)

    model_config = {"from_attributes": True}

File: frontend/src/types/document.ts
export type DocumentType = 'contract' | 'term' | 'material' | 'policy' | 'template' | 'other';
export type DocumentStatus = 'draft' | 'active' | 'archived';

export interface DocumentVersion {
  id: number;
  document_id: number;
  version_number: number;
  file_name: string | null;
  mime_type: string | null;
  file_size: number | null;
  external_url: string | null;
  notes: string | null;
  created_by_id: number | null;
  created_at: string;
}

export interface Document {
  id: number;
  title: string;
  document_type: DocumentType;
  description: string | null;
  status: DocumentStatus;
  course_id: number | null;
  class_offering_id: number | null;
  organization_id: number | null;
  student_id: number | null;
  uploaded_by_id: number | null;
  latest_version_number: number;
  is_signed: boolean;
  signed_at: string | null;
  signed_by: string | null;
  external_reference: string | null;
  created_at: string;
  updated_at: string;
  versions: DocumentVersion[];
}
